Cleanup: move NTLM status from defines to enum

kinkie opened the pull request:
NTLM status codes are defines and overlap somewhat.
Turn them into a C++ enum class.

yadij commented:
Looks reasonable, but why this change now?

kinkie commented:
I'm going after TODOs, starting from the simple ones, while waiting for the pending PRs to be reviewed

yadij commented:
PR #619 is waiting for your code update.

rousskov commented on src/auth/ntlm/SMB_LM/ntlm_smb_lm_auth.cc:
```diff
@@ -93,7 +93,7 @@ static unsigned char challenge[NTLM_NONCE_LEN];
        ntencoded_empty_pass[ENCODED_PASS_LEN];
 SMB_Handle_Type handle = NULL;
-int ntlm_errno;
+NtlmError ntlm_errno;
```
If possible, to avoid questions about this global being (not) shared with other translation units:

```suggestion
static NtlmError ntlm_errno;
```

I do not insist on this change.

kinkie replied:
It won't hurt so why not?

rousskov commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,24 +32,25 @@ extern "C" {
-#define NTLM_BAD_NTGROUP        2
-#define NTLM_BAD_REQUEST        3
-/* TODO: reduce the above codes down to one set non-overlapping. */
+enum class NtlmError : unsigned char
```
I suggest the following change to address the signed/unsigned/type conversion problems discussed in other change requests:

```suggestion
enum class NtlmError : int
```

Using `int` might also be faster, but that is not why I am suggesting this.

kinkie replied:
I think we can just not force the type then, and let the compiler do its magic. I believe that int is the default

rousskov commented:
I polished PR title for consistency sake and polished PR description punctuation a bit. Please adjust further as needed.

kinkie commented on src/auth/ntlm/SSPI/ntlm_sspi_auth.cc:
```diff
@@ -298,7 +298,7 @@ ntlm_check_auth(ntlm_authenticate * auth, char *user, char *domain, int auth_len
         domain[0] = '\0';
         x = ntlm_unpack_auth(auth, user, domain, auth_length);
```
I have rechecked all the defined constants and removed a few that had been left behind.

kinkie commented on src/auth/ntlm/SMB_LM/ntlm_smb_lm_auth.cc:
```diff
@@ -328,7 +328,7 @@ ntlm_check_auth(ntlm_authenticate * auth, int auth_length)
     debug("Login attempt had result %d\n", rv);
 
-    if (rv != NTLM_ERR_NONE) {  /* failed */
+    if (rv != NtlmError::None) {  /* failed */
```
Added an explicit switch statement to adapt between the two encodings

kinkie commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,24 +32,25 @@ extern "C" {
-#define NTLM_ERR_BAD_PROTOCOL     -1
-#define NTLM_ERR_NONE              0    /* aka. SMBLM_ERR_NONE */
-/* codes used by smb_lm helper */
-#define NTLM_ERR_SERVER            1    /* aka. SMBLM_ERR_SERVER   */
```
I don't think we need to keep them

rousskov replied:
I assume that you are implying that these `SMBLM_` constants were internal to Squid, and they are no longer used by any code we may need to integrate with (like old helpers or whatever). I consider this concern addressed and am resolving this change request.

rousskov commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,24 +32,25 @@ extern "C" {
-/* TODO: reduce the above codes down to one set non-overlapping. */
+enum class NtlmError : int
+{
+    None = 0,
+    ServerError = 1,
+    ProtocolError = 2,
+    LoginEror = 3,
+    UntrustedDomain = 4,
+    NtlmNotConnected = 5,
+    SspiError = 6,
+    BadNtGroup = 7,
+    BadRequest = 8,
+    InternalError = 128,
+    BlobError = 129,
+    BadProtocol = 130
```
The `None` value has to be zero because the current code relies on that (and because it is good practice). After your [clarification](https://github.com/squid-cache/squid/pull/676#discussion_r443688039), I see no need (and see harm in) explicitly numbering all other values, especially numbering them with a strange undocumented gap  after value `8`. Please remove explicit values other than zero.

kinkie replied:
ok

rousskov commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,24 +32,25 @@ extern "C" {
 
 /** String header. String data resides at the end of the request */
-typedef struct _strhdr {
+typedef struct _strhdr
```
You do not need to modify this code, but if you do modify it, then please drop `typedef` and `_strhdr`. Just use the regular C++ `class strhdr\n{\npublic: ...` template.

I recommend leaving these lines unchanged instead, to keep them consistent with other bad declarations in this file.

kinkie replied:
undone - it was unintentional. I checked for other formatting change, found none

rousskov commented on src/auth/ntlm/SSPI/ntlm_sspi_auth.cc:
```diff
@@ -573,20 +572,20 @@ manage_request()
             /* check against SSPI */
             int err = ntlm_check_auth((ntlm_authenticate *) decoded, user, domain, decodedLen);
             have_challenge = 0;
-            if (err != NTLM_ERR_NONE) {
+            if (err != NtlmError::None) {
```
C++11 [says](https://en.cppreference.com/w/cpp/language/enum) that "There are no implicit conversions from the values of a scoped enumerator to integral types". This is one of the primary differences between scoped and unscoped enumerators (by C++ design). AFAICT, this code requires such a conversion. My crude [example](https://cpp.godbolt.org/z/x5sGhu) fails to compile, as expected:

```
error: no match for 'operator!=' (operand types are 'int' and 'NtlmError')
     if (err != NtlmError::None)
```

I trust you have compiled this code. Can you explain why it works?

kinkie replied:
I have now understood: that helper has bitrotten, and it doesn't build.
I've tried building it with Wine libraries, but it fails with lots of redefined #defines in the socket layer, but also argument type errors

kinkie replied:
this specific issue can be fixed easily, as err is forcefully declared above but it's actually an NtlmError

kinkie replied:
auto type deduction fixes it. Resolved.

rousskov commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,24 +32,25 @@ extern "C" {
-#define NTLM_BAD_NTGROUP        2
-#define NTLM_BAD_REQUEST        3
-/* TODO: reduce the above codes down to one set non-overlapping. */
+enum class NtlmError : int
```
If an explicit underlying type is not needed, let's drop it.

```suggestion
enum class NtlmError
```

kinkie replied:
ok

kinkie commented:
I'm looking into setting up a mingw-based crosscompiler; let's see if it can unrot the windows-specific code.
@rousskov , if you agree I'd defer unrotting this helper to the windows specific effort

rousskov commented:
> if you agree I'd defer unrotting this helper to the windows specific effort

Sorry, I am not sure what that means for this PR. Please clarify; for example, are you proposing that this PR

1. knowingly damages some Windows-specific code further (because that code is known to be broken already)?
2. knowingly breaks some Windows-specific code (because you cannot easily test that code)?
3. modifies some Windows-specific code without testing it (because you cannot easily test that code)?
4. is put aside until you can test the affected Windows-specific code (so that you do not make things worse for Windows folks)?
5. does something else?

Please note that none of the above descriptions are meant to sound negative. I am just trying to understand what you want to do (and why) before I can answer your question in a meaningful way.

Needless to say, it would be nice to know what support tier we place Windows OS in, and what that tier means in terms of code breakage and PR testing, but I suspect we will not settle that issue by the time you need to get the answer to your question.

kinkie commented:
On Wed, 24 Jun 2020 at 23:08, Alex Rousskov <notifications@github.com>
wrote:

> if you agree I'd defer unrotting this helper to the windows specific effort
>
> Sorry, I am not sure what that means for this PR. Please clarify; for
> example, are you proposing that this PR
>
>    1. knowingly damages some Windows-specific code further (because that
>    code is known to be broken already)?
>    2. knowingly breaks some Windows-specific code (because you cannot
>    easily test that code)?
>    3. modifies some Windows-specific code without testing it (because you
>    cannot easily test that code)?
>    4. is put aside until you can test the affected Windows-specific code
>    (so that you do not make things worse for Windows folks)?
>    5. does something else?
>
>
The most similar is 4. The windows-only code is already broken, so these
changes to the windows-only ntlm-sspi helper can’t be tested. I’d like to
leave it broken for a short while, to avoid scope creep.

>
>    1.
>
> Please note that none of the above descriptions are meant to sound
> negative. I am just trying to understand what you want to do (and why)
> before I can answer your question in a meaningful way.
>
> Needless to say, it would be nice to know what support tier we place
> Windows OS in, and what that tier means in terms of code breakage and PR
> testing, but I suspect we will not settle that issue by the time you need
> to get the answer to your question.
>
I’m looking into adding mingw cross compiling on Linux for a windows
target. If that works it will integrate nearly in our build environment.
Next option is to have a windows build host, I’m looking into this using
the RackSpace donation.

Another option would be GitHub actions but that’s quite indirect so I’m not
sure if it’s worth the effort right now for this specific use case
-- 
@mobile

yadij commented:
FYI: Rafael at Diladele is on a full lockdown until COVID-19 was resolved in NL and so only supporting v3.5 for now due to some build issues with v4+. Once the lockdown is over I am hoping he can provide us access to a docker image we can integrate to the farm and collaborate better on build issues.

kinkie commented:
That would be great!
In the meantime, are you supporting my proposal or would you prefer a
different approach?
-- 
@mobile

yadij commented:
Since we know there are uncertain bugs in this code and unable currently to test it I would prefer not to be touching it at all. If this was triggered by your MinGW tests and found these defines to be a bug, that is fine. But cleanup just to remove a TODO is IMO needing ability to at least build test to accept - otherwise it may not actually be a *fix*.

rousskov commented:
> Alex: are you proposing that this PR ... (4) is put aside until you can test the affected Windows-specific code (so that you do not make things worse for Windows folks)?

> Francesco: The most similar is 4. The windows-only code is already broken, so these changes to the windows-only ntlm-sspi helper can’t be tested. I’d like to leave it broken for a short while, to avoid scope creep.

I fully support option 4 -- setting this PR aside until we can test it. Unfortunately, I do not think that is how you interpret that option, so I still do not know (or are not sure at all) what you are proposing. FWIW, "break the already broken code further" is option 1. There are many valid ways to proceed here -- I am not implying that option 4 is the only such option.

> to avoid scope creep

I see no scope creep. IMO: If we want to replace a group of related names used in Squid code, we should replace those names everywhere. If we touch tier-1 Squid code, we should test those changes before officially merging. If we cannot test tier-1 code, we should not polish it. Unfortunately, I do not know Windows' tier, and I am not sure there is consensus regarding these rules.

kinkie commented:
Suspending this PR until we have a windows test host to hash out kinks on

rousskov commented:
FTR: I replaced the `S-waiting-for-windows-testbed` label text with `S-waiting-for-QA` to slow down label proliferation and to continue naming a specific actor (to the extent possible) as the party responsible for ending the wait. I also changed the label color to match the rest of the `S-waiting-*` group.

yadij commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,21 +32,21 @@ extern "C" {
+    ProtocolError,
+    LoginEror,
+    UntrustedDomain,
+    NtlmNotConnected,
```
"Ntlm" prefix is redundant with the enum namespace. If anything it means DC not connected, but that does not need adding here either.

kinkie replied:
changed

rousskov commented:
> Alex, a year [ago](https://github.com/squid-cache/squid/pull/676#issuecomment-652409016): FTR: I replaced the S-waiting-for-windows-testbed label text with S-waiting-for-QA to slow down label proliferation and to continue naming a specific actor (to the extent possible) as the party responsible for ending the wait. I also changed the label color to match the rest of the S-waiting-* group.

> kinkie added the waiting-for-windows-qa label 3 minutes ago 

We can add platform labels if really needed, but I would not add a platform indication as other label suffixes to avoid label explosion. Once we cross a certain number of labels, they become very difficult for humans to use... I have now removed `waiting-for-windows-qa` label. If you decide to add platform label(s), please start/follow a common pattern like (`P-*`) to ease grouping of labels and label manipulation.

FWIW, our labeling goal should not be to describe the exact PR state, down to minute details (which would eventually require dealing with 1000 labels or 100-word labels). The goal should be to quickly tell the reader whether they should work on this PR (now) or skip it. The low-level details should be in the PR itself.

kinkie commented:
I added that to mark to myself that that PR is on a hiatus, as rightly
highlighted by Amos, we shouldn't touch what we can't test.
I'm hopefully going to have access to a windows machine soon, that should
help

On Fri, May 21, 2021 at 9:24 PM Alex Rousskov ***@***.***>
wrote:

> Alex, a year ago
> <https://github.com/squid-cache/squid/pull/676#issuecomment-652409016>:
> FTR: I replaced the S-waiting-for-windows-testbed label text with
> S-waiting-for-QA to slow down label proliferation and to continue naming a
> specific actor (to the extent possible) as the party responsible for ending
> the wait. I also changed the label color to match the rest of the
> S-waiting-* group.
>
> kinkie added the waiting-for-windows-qa label 3 minutes ago
>
> We can add platform labels if really needed, but I would not add a
> platform indication as other label suffixes to avoid label explosion. Once
> we cross a certain number of labels, they become very difficult for humans
> to use... I have now removed waiting-for-windows-qa label. If you decide
> to add platform label(s), please start/follow a common pattern like (P-*)
> to ease grouping of labels and label manipulation.
>
> FWIW, our labeling goal should not be to describe the exact PR state, down
> to minute details (which would eventually require dealing with 1000 labels
> or 100-word labels). The goal should be to quickly tell the reader whether
> they should work on this PR (now) or skip it. The low-level details should
> be in the PR itself.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/squid-cache/squid/pull/676#issuecomment-846193172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHPVDGOFTFCRJI3ZXH64KTTO2XOXANCNFSM4OD2A3SA>
> .
>


-- 
    Francesco

rousskov commented:
> I added that to mark to myself that that PR is on a hiatus

Sure, but we have to coordinate when it comes to Project-wide things such as PR labels. Unfortunately, GitHub does not support private labels.

FWIW, I recommend removing `S-waiting-for-author` if you are not going to work on this until we get a Windows machine - this PR is not really waiting on your action, it is waiting on QA infrastructure improvements, which `S-waiting-for-QA` already indicates. Without `S-waiting-for-author`, this PR will not be on your personal TODO list (if you compute that list using labels) _and_ others will not erroneously think that this PR is waiting for _your_ action.

And again, if you are sure that labeling a dominating or blocking PR platform/environment is very useful, then we should add a separate set of labels for that as well. My worry is that, in my experience, the more labeling categories one gets, the more difficult it becomes to label the PR correctly, and the fewer PRs will have correct/reliable labeling, defeating the primary purpose of the labels and making all labels a lot less useful.

yadij commented:
Aye, simpler is better. For S-waiting-* we go by group of stakeholders. With a COMMENT in the PR thread when unusual ones are set to indicate why.

kinkie commented:
I have now set up a Windows VM to test things, but not yet the tooling. I'll be able to unblock this soon

kinkie commented:
I've managed to successfully build the helper in mingw.
This should cover the requirement for QA:
```
$ do-build --parallel 10 --do-build --dir btbuild --mingw --continue --clean-before
[...]
$ ls -l btbuild/src/auth/ntlm/*/*.exe
-rwxr-xr-x 1 kinkie kinkie 58368 Dec 25 12:24 btbuild/src/auth/ntlm/SSPI/ntlm_sspi_auth.exe
```

rousskov commented on lib/ntlmauth/ntlmauth.h:
```diff
@@ -32,21 +32,21 @@ extern "C" {
-#define NTLM_BAD_NTGROUP        2
-#define NTLM_BAD_REQUEST        3
-/* TODO: reduce the above codes down to one set non-overlapping. */
+enum class NtlmError
```
This C++ code should not be surrounded by `extern "C" {}` but is.

I suspect this is not the only place where the updated header declarations have become C++ specific. Perhaps that wrapper should now be removed (or moved to become attached to a few function declarations that still need it)? To avoid a misunderstanding, I do not know the answer to that question. I only know that the current code looks self-contradictory to me.

kinkie replied:
That is a very good question. In Squid's codebase there are 53 files containing extern "C" definitions. Many of them are in compat/ , include/ and lib/ , but 12 are in src/ .

I'll make a note to look into whether these can be removed in a following PR

kinkie replied:
As for this specific change, I'm removing the extern linkage specifier. Thanks for catching this

rousskov commented on src/auth/ntlm/SMB_LM/ntlm_smb_lm_auth.cc:
```diff
@@ -317,33 +316,58 @@ ntlm_check_auth(ntlm_authenticate * auth, int auth_length)
+        ntlm_errno = NtlmError::LoginEror;
+        return nullptr;
+    default:
+        ntlm_errno = NtlmError::ServerError;
```
FWIW, this PR code hides/forgets the value returned by SMB_Logon_Server(). The PR description does not mention any functionality changes. I do not know whether this is a functionality change, but it may be worth double checking, especially if you are not testing this code (beyond building it).

kinkie replied:
Double checked. It is not a functionality change; the caller uses SMB_Get_Last_Error to dig deeper into the reasons for the failure if it so chooses.

rousskov reviewed (commented):
I cannot invest more time into this Windows-specific PR, but I left a couple of suggestions and am dismissing my previous review under the assumption that my earlier (grave) concerns have been addressed.

yadij commented on src/auth/ntlm/SSPI/ntlm_sspi_auth.cc:
```diff
@@ -164,178 +164,68 @@ Valid_Group(char *UserName, char *Group)
     return result;
 }
 
-char * AllocStrFromLSAStr(LSA_UNICODE_STRING LsaStr)
```
Please separate these dead-code removal to a different PR than the enum refactor.

kinkie replied:
Done. PR #1627

yadij commented on src/auth/ntlm/SMB_LM/ntlm_smb_lm_auth.cc:
```diff
@@ -317,33 +316,58 @@ ntlm_check_auth(ntlm_authenticate * auth, int auth_length)
+    case SMBlibE_Remote:
+        ntlm_errno = NtlmError::ServerError;
+        return nullptr;
+    case SMBlibE_GuestOnly:
+        ntlm_errno = NtlmError::LoginEror;
+        return nullptr;
```
Please move this case up just below `SMBlibE_Success` so the common cases occur first in order of expected occurance.  This is a performance optimization, I do not insist.

yadij reviewed (dismissed):
LGTM, provided it builds after this latest merge.

kinkie commented:
I’ll hand check it and attach evidence, this code is not built as part of
the normal procedures

@mobile


On Sat, 30 Dec 2023 at 13:13, Amos Jeffries ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> LGTM, provided it builds after this latest merge.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/squid-cache/squid/pull/676#pullrequestreview-1799553046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHPVDFLQNF6V5NPMMLTRM3YMAAOPAVCNFSM4OD2A3SKU5DIOJSWCZC7NNSXTPCQOVWGYUTFOF2WK43UKJSXM2LFO45TCNZZHE2TKMZQGQ3A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kinkie commented:
Build evidence: 
$ do-build --parallel 10 --do-build --dir btbuild --log ../build.log --mingw --continue --clean-before
[...]
$ file btbuild/src/auth/*/*/*.exe
btbuild/src/auth/basic/fake/basic_fake_auth.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows
btbuild/src/auth/basic/SSPI/basic_sspi_auth.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows
btbuild/src/auth/ntlm/SSPI/ntlm_sspi_auth.exe:   PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows

kinkie commented:
@yadij , merging from master has voided your previous approval; could you re-approve?

kinkie commented:
ok to test

kinkie commented:
@rousskov do you mind stamping this PR? As you have been requested to review, it can't move forward

rousskov commented:
> @rousskov do you mind stamping this PR? As you have been requested to review, it can't move forward

My status has not changed since https://github.com/squid-cache/squid/pull/676#pullrequestreview-1796024275: I cannot invest more time into this Windows-specific PR. I am declining this new opportunity to review it under the old assumption that my earlier (grave) concerns have been addressed.

kinkie commented:
> > @rousskov do you mind stamping this PR? As you have been requested to review, it can't move forward
> 
> My status has not changed since [#676 (review)](https://github.com/squid-cache/squid/pull/676#pullrequestreview-1796024275): I cannot invest more time into this Windows-specific PR. I am declining this new opportunity to review it under the old assumption that my earlier (grave) concerns have been addressed.

that's okay, I'm not asking that you endorse code you can't test. Github was showing that you had been requested to review and pending that was not allowing the merge to proceed. That has now changed and the reviews by myself and Amos can allow the process to continue
